scripts/test.sh

```diff
@@ -52,14 +53,15 @@ download_server() {
 
 start_server() {
   echo "test.sh: starting ROS"
-  #disabled ROS logging
-  # sh ./object-server-for-testing/start-object-server.command &> /dev/null &
-
-  #enabled ROS logging
-  #sh ./object-server-for-testing/start-object-server.command &
-  export ROS_SKIP_PROMPTS=true &&  ./node_modules/.bin/ros start --data realm-object-server-data &
-  SERVER_PID=$!
+  if [ -z "${SYNC_WORKER_FEATURE_TOKEN}" ]; then
+      die "SYNC_WORKER_FEATURE_TOKEN must be set to run tests."
+  fi
+  mkdir -p "$(pwd)/build"
+  ros_log_temp="$(pwd)/build/ros_out.txt"
+  ROS_SKIP_PROMPTS=true ./node_modules/.bin/ros start --data realm-object-server-data 2>&1 | tee $ros_log_temp &
+  SERVER_PID=$(jobs -l | grep node_modules/.bin/ros | cut -f2 -d" ")
```

Comment:
Will this work if multiple ROS instances are running on the same CI slave?

Reply:
A single CI worker can't run multiple jobs at once to begin with.

Reply:
Some of the CI nodes have multiple executors? Also - what if this script is running on a developers machine which has another instance of ROS running besides running these tests?

Reply:
On Linux the CI jobs are all run inside docker containers. A developer running their own separate copy of ROS when they call test.sh has never worked, but not because of this line. `jobs` gives a list of all children spawned by this bash script.

Reply:
I remember being able to run the tests sing `test.sh` on my local machine once and I think we should strive for that. Alternatively we should make it clear that we don't intent that `test.sh` should be used locally to avoid confusion and wasted time tracking down bugs arising from assumptions like this.

Reply:
You certainly can run test.sh locally (that's part of the point of putting the logic in a shell script rather than in Jenkins), but if you want to run the tests when you already have ROS running it seems obvious that you should just run the tests directly rather than going through the script that also spins up the external deps?

If you forget that you have ROS running it'll just fail to start up with a port in use error. That's the same thing it's done ever since the sync tests were first added here, and I'm not sure why it's really relevant to this PR.

Reply:
In case it's unclear, there's two general ways that I go about running realm-js tests locally:

1) When I want to run exactly what will be run on CI, I use test.sh. This will fail if I already have a separate copy of ROS running on the same port as it tries to use for its copy of ROS. This is the behavior I want, as anything else would not actually be testing the same thing as what runs on CI.
2) When I'm actively developing things, I fire up a copy of ROS (and metro bundler if I'm touching react native), open the Xcode project and run the tests from that without touching test.sh.

Nothing about any of this is changing in this PR. The only thing different is that test.sh used to not actually wait for the copy of ROS it launched to start up, so the very first test would sometimes fail.

Reply:
Thanks for the clarification.

An alternative to grepping from the jobs could be `& echo $!` to get the PID of the ROS server:

```bash
SERVER_PID=$(ROS_SKIP_PROMPTS=true ./node_modules/.bin/ros start --data realm-object-server-data 2>&1 | tee $ros_log_temp & echo $!)
```

Just a suggestion :)

Reply:
That gives the pid of `tee`, not the node process.

What is the problem with grepping `jobs` that needs to be fixed?

Reply:
It just feels brittle to me. Grepping jobs might return the PID of another server is running on the developers machine (which might not be using the same ports - hence won't fail the tests from starting the server as you commented). If that occurs, it's one of those issues that will be super hard to debug as it would be unexpected from just running a test script. But if it's hard to find a more robust way of getting the PID, we shouldn't waste cycles or hold back this PR - just wanted to point out the potential issue.

Reply:
`jobs` returns a list of processes started by the current shell. It will not return the PID of another server running on the developer's machine.

Reply:
Oh - I didn't realize that.